Objects/typeobject.c

```diff
@@ -6772,6 +6759,31 @@ type_ready_set_new(PyTypeObject *type)
     return 0;
 }
 
+static int
+type_ready_managed_dict(PyTypeObject *type)
+{
+    if (!(type->tp_flags & Py_TPFLAGS_MANAGED_DICT)) {
+        return 0;
+    }
+    if (!(type->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
+        PyErr_Format(PyExc_SystemError,
+                     "type %s has the Py_TPFLAGS_MANAGED_DICT flag "
+                     "but not Py_TPFLAGS_HEAPTYPE flag.",
```

Comment:
I noticed that and I'm surprised that PyType_Ready() can actually set the flag on a static type which inherits from a heap type in the inherit_special() function. It's unclear to me what happens in this case. Likely a crash. Hopefully, this new check should prevent the crash!